File: 面试题35-复杂链表的复制.py
class Solution:
    def copyRandomList(self, head: 'Node') -> 'Node':
        if not head:
            return None
        
        mydict = {None: None}
        cur_node = head

        while cur_node:
            if not mydict.get(cur_node):
                mydict[cur_node] = Node(cur_node.val)
            if cur_node.next and not mydict.get(cur_node.next):
                mydict[cur_node.next] = Node(cur_node.next.val)
            if cur_node.random and not mydict.get(cur_node.random):
                mydict[cur_node.random] = Node(cur_node.random.val)

            mydict[cur_node].next = mydict[cur_node.next]
            mydict[cur_node].random = mydict[cur_node.random]   

            cur_node = cur_node.next         

        return mydict[head]

File: 面试题35-复杂链表的复制.cpp
class Solution {
public:
    Node* copyRandomList(Node* head) {
        if(!head)
            return NULL;

        Node* cur_node = head;
        while(cur_node){
            Node* new_node = new Node(cur_node->val);
            new_node->next = cur_node->next;
            cur_node->next = new_node;
            cur_node = cur_node->next->next;
        }

        cur_node = head;
        while(cur_node){
            cur_node->next->random = cur_node->random ? cur_node->random->next : NULL;
            cur_node = cur_node->next->next;
        }

        Node* new_head = head->next;
        Node* cur_new_node = new_head;
        cur_node = head;

        while(cur_node){
            cur_node->next = cur_new_node->next;
            cur_new_node->next = cur_node->next ? cur_node->next->next : NULL;
            cur_node = cur_node->next;
            cur_new_node = cur_new_node->next;
        }

        return new_head;
    }
};
